src/test/java/org/killbill/billing/plugin/vertex/VertexTaxCalculatorTest.java

```diff
@@ -231,6 +231,95 @@ public void testTaxDescription() throws Exception {
+        result = vertexTaxCalculator.compute(account, invoice, true, Collections.emptyList(), tenantContext);
+        assertNull(result.get(0).getItemDetails());
+        checkTaxItemFields(result.get(0));
+    }
+
+    @Test(groups = "fast")
+    public void testComputeWhenTaxEffectiveRateIsNotPresented() throws Exception {
+        //given
+        given(invoice.getInvoiceItems()).willReturn(Collections.singletonList(taxableInvoiceItem));
+        given(taxResponse.getData()).willReturn(apiResponseData);
+        final TaxesType taxesType = new TaxesType();
+        taxesType.setCalculatedTax(MOCK_TAX_AMOUNT_1_01);
+        given(responseLineItem.getTaxes()).willReturn(Collections.singletonList(taxesType));
+
+        //given tax effective rate is not present
+        taxesType.setEffectiveRate(null);
+
+        //then it persisted in item details invoice item field
+        List<InvoiceItem> result = vertexTaxCalculator.compute(account, invoice, true, Collections.emptyList(), tenantContext);
+        assertEquals(1, result.size());
+        assertNull(result.get(0).getItemDetails());
+        checkTaxItemFields(result.get(0));
+    }
+
+    @Test(groups = "fast")
+    public void testComputeWhenEffectiveTaxRateZero() throws Exception {
+        //given
+        given(invoice.getInvoiceItems()).willReturn(Collections.singletonList(taxableInvoiceItem));
+        given(taxResponse.getData()).willReturn(apiResponseData);
+        final TaxesType taxesType = new TaxesType();
+        taxesType.setCalculatedTax(0d);
+        given(responseLineItem.getTaxes()).willReturn(Collections.singletonList(taxesType));
+
+        //given tax effective rate is present
+        taxesType.setEffectiveRate(0d);
+
+        //then it persisted in item details invoice item field
+        List<InvoiceItem> result = vertexTaxCalculator.compute(account, invoice, true, Collections.emptyList(), tenantContext);
+        assertEquals(0, result.size());
+    }
+
+    @Test(groups = "fast")
+    public void testComputeWhenEffectiveTaxRateZeroButAmountIsNot() throws Exception {
+        //given
+        given(invoice.getInvoiceItems()).willReturn(Collections.singletonList(taxableInvoiceItem));
+        given(taxResponse.getData()).willReturn(apiResponseData);
+        final TaxesType taxesType = new TaxesType();
+        taxesType.setCalculatedTax(MOCK_TAX_AMOUNT_1_01);
+        given(responseLineItem.getTaxes()).willReturn(Collections.singletonList(taxesType));
+
+        //given tax effective rate is present
+        taxesType.setEffectiveRate(0d);
+
+        //then it persisted in item details invoice item field
+        List<InvoiceItem> result = vertexTaxCalculator.compute(account, invoice, true, Collections.emptyList(), tenantContext);
+        assertEquals(1, result.size());
+        assertEquals("{\"taxRate\":0.0}", result.get(0).getItemDetails());
+        checkTaxItemFields(result.get(0));
+    }
+
+    private void checkTaxItemFields(final InvoiceItem taxItem) {
```

Comment:
To my understanding, below are the practical cases -

1. InvoiceItem has 0 tax - No tax item should be generated
2. InvoiceItem has non-zero tax - taxItem with ItemDetails is generated.

Other scenario where tax-rate is null should not happen if tax-amount != 0

Reply:
I think the key word is "should not happen". But it can, doesn't it?

Reply:
discussed